File: src/main/python/util/LinearRegression.py
#
#   Copyright 2008-2015 Semantic Discovery, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
import math

class LinearRegression:
    '''
    A low-memory helper class to collect (x,y) samples and compute the
    linear regression.
    '''

    def __init__(self):
        self._n = 0
        self._x_sum = 0.0
        self._y_sum = 0.0
        self._xy_sum = 0.0
        self._xx_sum = 0.0
        self._yy_sum = 0.0

        self._m = None
        self._b = None

    @property
    def n(self):
        return self._n

    @property
    def x_sum(self):
        return self._x_sum

    @property
    def y_sum(self):
        return self._y_sum

    @property
    def xy_sum(self):
        return self._xy_sum

    @property
    def xx_sum(self):
        return self._xx_sum

    @property
    def yy_sum(self):
        return self._yy_sum

    @property
    def m(self):
        return self.getM()

    @property
    def b(self):
        return self.getB()

    def getY(self, x):
        return self.getM() * x + self.getB()

    def add(self, x, y):
        self._m = None
        self._b = None
        self._n += 1
        self._x_sum += x
        self._y_sum += y
        self._xy_sum += (x * y)
        self._xx_sum += (x * x)
        self._yy_sum += (y * y)

    def __str__(self):
        return 'y = %.4f x + %.4f' % (self.getM(), self.getB())

    def getM(self):
        if (self._m is None):
            self._computeRegression()
        return self._m

    def getB(self):
        if (self._b is None):
            self._computeRegression()
        return self._b

    def _computeRegression(self):
        self._m = 0.0
        self._b = 0.0

        denominator = self._n * self._xx_sum - self._x_sum * self._x_sum
        if (denominator != 0):
            m_numerator = self._n * self._xy_sum - self._x_sum * self._y_sum
            b_numerator = self._y_sum * self._xx_sum - self._x_sum * self._xy_sum

            self._m = m_numerator / denominator
            self._b = b_numerator / denominator

File: src/main/java/org/sd/util/LinearRegression.java
/*
    Copyright 2009-2017 Semantic Discovery, Inc. (www.semanticdiscovery.com)

    This file is part of the Semantic Discovery Toolkit.

    The Semantic Discovery Toolkit is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    The Semantic Discovery Toolkit is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with The Semantic Discovery Toolkit.  If not, see <http://www.gnu.org/licenses/>.
*/
package org.sd.util;


/**
 * A helper class to collect (x,y) samples and compute the linear regression.
 *
 * @author Spence Koehler
 */
public class LinearRegression {

  private long n;
  private double x_sum;
  private double y_sum;
  private double xy_sum;
  private double xx_sum;
  private double yy_sum;

  private Double _m;
  private Double _b;

  /**
   * Default constructor.
   */
  public LinearRegression() {
    this.n = 0L;
    this.x_sum = 0.0;
    this.y_sum = 0.0;
    this.xy_sum = 0.0;
    this.xx_sum = 0.0;
    this.yy_sum = 0.0;

    this._m = null;
    this._b = null;
  }

  public double getY(double x) {
    return getM() * x + getB();
  }

  public double getM() {
    if (this._m == null) {
      computeRegression();
    }
    return this._m;
  }

  public double getB() {
    if (this._b == null) {
      computeRegression();
    }
    return this._b;
  }

  public String toString() {
    return String.format("y = %.4f x + %.4f", getM(), getB());
  }

  /**
   * Add a sampled value to this instance.
   */
  public void add(double x, double y) {
    ++n;
    x_sum += x;
    y_sum += y;
    xy_sum += (x * y);
    xx_sum += (x * x);
    yy_sum += (y * y);

    _m = null;
    _b = null;
  }

  /**
   * Get the number of sampled values.
   */
  public long getN() {
    return n;
  }

  /**
   * Get the sum of all "x" values.
   */
  public double getXSum() {
    return x_sum;
  }

  /**
   * Get the sum of all "y" values.
   */
  public double getYSum() {
    return y_sum;
  }
  
  /**
   * Get the sum of all "xy" values.
   */
  public double getXYSum() {
    return xy_sum;
  }

  /**
   * Get the sum of all "xx" values.
   */
  public double getXXSum() {
    return xx_sum;
  }

  /**
   * Get the sum of all "yy" values.
   */
  public double getYYSum() {
    return yy_sum;
  }

  private final void computeRegression() {
    this._m = 0.0;
    this._b = 0.0;

    final double denominator = n * xx_sum - x_sum * x_sum;
    if (denominator != 0) {
      final double m_numerator = n * xy_sum - x_sum * y_sum;
      final double b_numerator = y_sum * xx_sum - x_sum * xy_sum;

      this._m = m_numerator / denominator;
      this._b = b_numerator / denominator;
    }
  }
}
